### PR TITLE
Refactor/client types

### PR DESCRIPTION
**Pull request by Awk34**

### CODE REVIEW

**Comment by Awk34** [comment]
@Koslun these changes should fix the TS errors *and* work in both TS/Flow

**Comment by Koslun** [comment]
@Awk34 Take it we're skipping more specific typings from definition files that would only apply to TS to avoid having *a lot* of ejs if's and thus more maintainable templates? Especially since many will be useless with Angular 2 with more maintainable templates also being a plus when we're upgrading.

**Comment by Awk34** [comment]
@Koslun I'm just trying to fix all the TS errors that people are seeing when building the app for the first time. I also want to try to keep any type definitions able to be used with both Flow & TS. Is that reasonable?

Also, is this good to merge?

**Comment by Koslun** [comment]
> @Koslun I'm just trying to fix all the TS errors that people are seeing when building the app for the first time. I also want to try to keep any type definitions able to be used with both Flow & TS. Is that reasonable?

As I said. Appreciate that restricting type definitions to what works in both flow and TS enables us to avoid having either an extra template file or just a lot of if-statements for when it's TS. Also realize that most of TS-specific logic will be useless with Angular 1 anyway. With Angular 2 using external type definitions will also be needed to a much lesser extent as the Angular 2 libraries are already written in TS. So figure that fully typed TS and Flow code will be a lot more similar then.

So based on that premise I think it's fully reasonable and like the strategy. 

If anything I'm more skeptical of Flow itself as it does not currently support definition files and I know no libraries written with Flow, would think React is but cannot find anything to indicate that. Can however somewhat understand wanting to rather work with Babel and thus using Flow with that. And as it's already implemented in this generator I can see the benefits of it.

> Also, is this good to merge?
Pretty much, seems to work on my end. Might want to fix two final errors:

```
[default] project/client/components/auth/auth.service.ts:137:14
    Supplied parameters do not match any signature of call target.
[default] project/client/components/auth/auth.service.ts:162:14
    Supplied parameters do not match any signature of call target.
```
Looking into it now.

**Comment by Awk34** [comment]
> would think React is

Yeah, React uses Flow.

**Comment by Koslun** [comment]
What I don't then get is why they don't clearly state that here: https://facebook.github.io/react/docs/language-tooling.html#flow , or even here: https://flowtype.org/docs/react.html#_. Feels like a pretty meaningful feature and a very strong insentive to use Flow when you develop React applications. Not mentioned on React's wikipedia page either.

**Comment by Awk34** [comment]
@Koslun ¯\\\_(ツ)\_/¯

They have TypeScript typings, as well as test against TypeScript, CoffeeScript

**Comment by Koslun** [comment]
Shouldn't have changed any behavior just more explicitly sending in `undefined`, could also just send the functions in the `then` blocks too but opted to just add small note to documentation and change nothing really.

File: templates/app/client/components/auth(auth)/auth.service.js [hunk 1]
```diff
@@ -100,7 +100,7 @@ export function AuthService($location, $http, $cookies, $q, appConfig, Util, Use
     /**
      * Gets all available info on a user
      *
-     * @param  {Function} [callback] - funciton(user)
+     * @param  {Function} [callback] - optional, function(user)
```
Comment (Awk34) [inline templates/app/client/components/auth(auth)/auth.service.js]: "optional, " is redundant here

File: templates/app/client/components/auth(auth)/auth.service.js [hunk 2]
```diff
@@ -100,7 +100,7 @@ export function AuthService($location, $http, $cookies, $q, appConfig, Util, Use
...
      * @return {Promise}
      */
     getCurrentUser(callback) {
```
Comment (Awk34) [inline templates/app/client/components/auth(auth)/auth.service.js]: wouldn't the fix be to say `callback?: Function`?

Comment (Koslun) [inline templates/app/client/components/auth(auth)/auth.service.js]: Yeah that's a more informative type definition for it. Was not sure it was valid Flow syntax as well but it does appear to be. Examples or documentation of each for reference: [Flow documentation](https://flowtype.org/docs/functions.html) [TS example](http://www.typescriptlang.org/play/#src=export%20interface%20INotificationService%20%7B%0A%20%20%20%20error(message%3A%20string%2C%20title%3F%3A%20string%2C%20autoHideAfter%3F%20%3A%20number)%3B%0A%7D%0A%0Aclass%20X%20%7B%0A%09error(message%3A%20string%2C%20title%3F%3A%20string%2C%20autoHideAfter%3F%3A%20number)%20%7B%0A%09%09console.log(message%2C%20title%2C%20autoHideAfter)%3B%0A%09%7D%0A%7D%0A%0Anew%20X().error(%22hi%20there%22%2C%20undefined%2C%201000)%3B).

Should probably change all the other async functions then in auth.service as I kinda only copied documentation from them while all of them use the same `Util.safeCb` to make it optional. 

Will just do it now.

**Comment by Koslun** [comment]
Think the fail is unrelated, tried scaling back the changes and getting the same errors when running the tests locally.

**Comment by Awk34** [comment]
Sequelize likes to :shit: itself quite often

**Pull request merged**